packages/core/src/http-client/csrf-token-header.ts

```diff
@@ -55,26 +60,52 @@ function makeCsrfRequest<T extends HttpRequestConfig>(
 
   // The S/4 does a redirect if the CSRF token is fetched in case the '/' is not in the URL.
   // TODO: remove once https://github.com/axios/axios/issues/3369 is really fixed. Issue is closed but problem stays.
-  return executeHttpRequest(destination, appendSlash(axiosConfig))
+  const requestConfigWithTrailingSlash = appendSlash(axiosConfig);
+  return executeHttpRequest(destination, requestConfigWithTrailingSlash)
     .then(response => response.headers)
-    .catch(e1 => {
-      logger.error(
+    .catch(error1 => {
+      if (hasCsrfToken(error1)) {
+        return getResponseHeadersFromAxiosError(error1);
+      }
+      logger.warn(
         new ErrorWithCause(
-          'Initial try to fetch CSRF token failed - retry without slash at ',
-          e1
+          `First attempt to fetch CSRF token failed with the url: ${requestConfigWithTrailingSlash.url}. Retrying without trailing slash.`,
+          error1
         )
       );
-      return executeHttpRequest(destination, removeSlash(axiosConfig))
+      const requestConfigWithOutTrailingSlash = removeSlash(axiosConfig);
+      return executeHttpRequest(destination, requestConfigWithOutTrailingSlash)
         .then(response => response.headers)
-        .catch(e2 => {
-          throw new ErrorWithCause(
-            'Also second try to fetch SRF token failed - No CSRF token fetched.',
-            e2
+        .catch(error2 => {
+          if (hasCsrfToken(error2)) {
+            return getResponseHeadersFromAxiosError(error2);
+          }
+          logger.warn(
+            new ErrorWithCause(
+              `Second attempt to fetch CSRF token failed with the url: ${requestConfigWithOutTrailingSlash.url}. No CSRF token fetched.`,
+              error2
+            )
           );
+          // todo suggest to disable csrf token handling when the api is implemented
+          return {};
         });
     });
 }
 
+// In general, the valid csrf token head can also be found in a non-2xx request.
+// For a non-2xx response, axios will throw an error so we have to check whether the error contains the token.
+function hasCsrfToken(error: any): boolean {
+  return error.isAxiosError && hasCsrfTokenInAxiosError(error);
+}
+
+function hasCsrfTokenInAxiosError(error: AxiosError): boolean {
+  return error.response?.headers?.['x-csrf-token'];
+}
+
+function getResponseHeadersFromAxiosError(error: AxiosError) {
+  return error.response!.headers;
+}
+
```

Comment:
This is still very dependent on axios and I don't see why it needs to be.
Consider the code below. It does the same thing but there is no notion of axios.
Also, it makes the flow a little less redundant (accessing the headers and then checking it once).
```ts
function hasCsrfToken(headers: Record<string, any>): boolean {
  return !!headers.['x-csrf-token'];
}

// Non-2xx responses can contain valid csrf tokens in their headers.
function getResponseHeadersFromError(error: any): Record<string, any> {
  return error.response?.headers || {};
}

// Usage
const headers = getResponseHeadersFromError(errorX);
if (hasCsrfToken(headers)) {
  return headers;
}
...
```